Move insertionPoint state to block-editor store/rename existing insertionPoint to insertionCue

jeryj opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
In the block editor store:
- Change `insertionPoint` reducer to `insertionCue` (as it is where the blue line shows up _not_ where insertion is guaranteed to happen).
- This opens us `insertionPoint` to be used to manually pass an insertion point (such as when clicking an inline inserter button like the blue zoom out inserter buttons)
- Add `setInsertionPoint` action
- Adds private `getInsertionPoint` selector 

In the editor store:
- `setIsInserterOpened` dispatches `setInsertionPoint` to the block editor store
- Rename editor selector `getInsertionPoint` to `getInserter` to more clearly communicate this is the state of the inserter, not solely about the insertion point. 
## Why?/How
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Now that the inserter can remain open while interacting with the block canvas, it is expected that when you select a new block, the insertion will occur after the selected block. 

The insertion point is handled by the editor package via `setIsInserterOpened( value )`, which can take an object to open the inserter to a certain tab panel and also provide the place where the insertion should happen. We want to change this insertion point when selecting a new block on the canvas. Block selection is handled via the block editor package. Since the block editor should not know about the editor, and we can't rely on the `SELECTION_CHANGE` action to clear the insertion point within the editor store, I've moved the insertion point data to be within the block editor package. 

In order to clear up confusion, I've also renamed some state and selectors:
- Block editor `insertionPoint` state becomes `insertionCue`, as this more accurately names it. It was not the insertion point, but the place where we want to show the blue line on the canvas.
- This opens up `insertionPoint` to be used for manually setting an insertion point, if needed, such as when opening the inserter via one of the inline inserter buttons.
- In the editor package, `getInsertionPoint` is renamed to `getInserter`, as this was a point of confusion. It is only the insertion point if the inserter is open and becomes stale data as soon as another interaction on the canvas happens. This is about the sate of the inserter when it is opened, and if it should still be open. `getInserter` is a better name, and it could still be better. `getInserterPanel` maybe?

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
- Renames `state.insertionPoint` to `state.insertionCue` in the block editor store and related action/reducers.
- Adds `state.insertionPoint` to store when an insertion point is manually passed.
- Emit the private `setInsertionPoint` action in the block editor store when `setIsInserterOpen` has an root index and insertion point set.
- Clear `insertionPoint` state on `SELECT_BLOCK` 
- Renames private editor selector `getInsertionPoint` to `getInserter` to more accurately communicate that this is the state the inserter initialized to rather than the final insertion point.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->
### For pattern insertion points in Zoom Out
1. Open the Site Editor
2. Enter zoom out mode (this isn't necessary but it makes it much easier to test) - you can enter zoom out by selecting a pattern category in the inserter or by using the 50% desktop option in the device preview dropdown)
4. Click one on the [+] in the canvas, to set an insertion point. Notice that a wide blue line appears between the two blocks.
5. Select another block, or clear block selection. Notice that that wide blue line disappears.
6. Now select a pattern in the canvas. This pattern should be inserted after the newly selected block, not at the place where the old insertion point was.
7. Create an insertion point by clicking the [+]. Hover a pattern in the inserter. You should still see the wide blue line in the canvas.
9. Now click the pattern. It should be added to the place that the insertion point indicator was.

### Default Editing
- Use the various in-between inserters and inserters in both group and root level contexts (like a new post). 
- See if insertion always happens where you would expect it to
- If it is unexpected, check to see if it's different than trunk

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

Mamaduka commented:
I was also trying to find the right solution for this issue and came to a similar conclusion. We should probably fully revert #64048 and work on an alternative fix for #63866.

Why revert? Zoom-out mode is still experimental (behind the settings flag); experimental features should hinder stable ones.

jeryj committed:
Use index instead of insertionIndex

jeryj committed:
Revert "Use index instead of insertionIndex"

This reverts commit df849a92860f7e9b75f56153627370017c005221.

jeryj commented:
@Mamaduka - I'm fine reverting https://github.com/WordPress/gutenberg/pull/64048 as well. I've been looking at this most of the day and feel like there's a deeper issue at play. 

The insertion point is bound to the editor, and gets set when the inserter opens. The idea of the inserter staying open and the insertion point changing wasn’t a thing when this was designed, because the inserter was _supposed to close and reset when focus was moved out of it_. Now that it stays open and you can interact with the canvas, we have this muddling of states where the insertion point can change while the inserter is open.

jeryj committed:
Move/Add inserterInsertionPoint to block editor

jeryj commented:
I worked more on this today to move the insertion point into the block editor package. It's working correctly for the zoom out inserters. I really dislike the naming of all the things as it's so hard to untangle which insertion point means what :/

Mamaduka commented:
cc @WordPress/gutenberg-core

jeryj committed:
Update the docs of setIsInserterOpened

jeryj commented on packages/editor/src/store/actions.js:
```diff
@@ -726,6 +726,10 @@ export function removeEditorPanel( panelName ) {
  *                                              use an object.
  * @param {string}         value.rootClientId   The root client ID to insert at.
  * @param {number}         value.insertionIndex The index to insert at.
+ * @param {string}         value.filterValue    A query to filter the inserter results.
+ * @param {Function}       value.onSelect       A callback when an item is selected.
+ * @param {string}         value.tab            The tab to open in the inserter.
+ * @param {string}         value.category       The category to initialize in the inserter.
```
These are not new, just adding the documentation for things already in use.

jeryj commented on docs/reference-guides/data/data-core-editor.md:
```diff
@@ -1422,6 +1422,10 @@ _Parameters_
 -   _value_ `boolean|Object`: Whether the inserter should be opened (true) or closed (false). To specify an insertion point, use an object.
 -   _value.rootClientId_ `string`: The root client ID to insert at.
 -   _value.insertionIndex_ `number`: The index to insert at.
+-   _value.filterValue_ `string`: A query to filter the inserter results.
+-   _value.onSelect_ `Function`: A callback when an item is selected.
+-   _value.tab_ `string`: The tab to open in the inserter.
+-   _value.category_ `string`: The category to initialize in the inserter.
```
These are not new, just adding the documentation for things already in use.

jeryj committed:
Remove unnecessary props since they won't be set

jeryj commented on packages/block-editor/src/components/inserter/hooks/use-insertion-point.js:
```diff
@@ -79,18 +79,26 @@ function useInsertionPoint( {
 			} else if ( clientId ) {
 				// Insert after a specific client ID.
 				_destinationIndex = getBlockIndex( clientId );
+			} else if (
+				insertionPoint?.insertionIndex &&
+				insertionPoint?.rootClientId
+			) {
+				_destinationRootClientId = insertionPoint.rootClientId;
+				_destinationIndex = insertionPoint.insertionIndex;
```
Putting this after the `insertionIndex` or `clientId ` to preserve the existing API. This lets anyone who passes these values in via the props to still have full control.

jeryj commented on packages/block-editor/src/store/private-actions.js:
```diff
@@ -359,6 +359,20 @@ export function expandBlock( clientId ) {
+ *
+ * @return {Object} Action object.
+ */
+export function setInserterInsertionPoint( value ) {
+	return {
+		type: 'SET_INSERTER_INSERTION_POINT',
+		value,
+	};
+}
+
```
I don't like this naming, but it's at least clear 🤷🏻  I'm concerned that this assumes only one instance of the inserter, while the editor package way passes it in. That said, it's still possible to manually pass the inserter the root block and index and it will be respected.

jeryj commented on packages/block-editor/src/store/reducer.js:
```diff
@@ -2087,6 +2087,25 @@ export function hoveredBlockClientId( state = false, action ) {
+	switch ( action.type ) {
+		case 'SET_INSERTER_INSERTION_POINT':
+			return action.value;
+		case 'SELECT_BLOCK':
+			return null;
```
Clear the insertion point after a new block selection.

scruffian replied:
In my testing, the vertical displacement is still visible after selecting a different block.

jeryj committed:
Discourage use of rootIndex and insertionIndex on setIsInserterOpened

jeryj committed:
Remove note

jeryj committed:
Use setInserterInsertionPoint in quick inserter

Mamaduka commented:
@jeryj, what do you think about reverting #64048 in a separate PR, while we continue working on a fix here?

jeryj commented:
@Mamaduka - yes! Of course. Here's a revert PR: https://github.com/WordPress/gutenberg/pull/65208

jeryj committed:
Insert in correct location for root insertions

jeryj committed:
Add store tests and improve docs and default state

Mamaduka commented:
[Performance charts](https://www.codevitals.run/project/gutenberg/inserterHover) after the revert PR was merged. We should monitor "Inserter Hovering Items" results while working on the fix.

## Screenshot
![CleanShot 2024-09-11 at 12 20 30](https://github.com/user-attachments/assets/b2a414a2-4896-4e17-8254-c5e45df046d4)

kevin940726 reviewed (commented):
I noticed something weird that I can't reproduce on trunk. I'm not sure if it's related to this PR nor what's going on. The testing environment is Playground.


https://github.com/user-attachments/assets/d533f08d-bd13-4135-9f86-2f8d0181bd68

Notice how sometimes selecting the pattern will insert it at the wrong place. And some other times hover over sections will rerender the patterns inserter.

getdave commented on packages/block-editor/src/components/inserter/hooks/use-insertion-point.js:
```diff
@@ -79,15 +79,21 @@ function useInsertionPoint( {
 				// Insert into a specific index.
 				_destinationIndex = insertionIndex;
+			} else if ( insertionPoint?.insertionIndex ) {
+				_destinationRootClientId = insertionPoint?.rootClientId
+					? insertionPoint.rootClientId
+					: rootClientId;
+				_destinationIndex = insertionPoint.insertionIndex;
```
The other conditions have comments to explain them. Could we add one here?

getdave commented on packages/block-editor/src/components/inserter/quick-inserter.js:
```diff
@@ -82,15 +83,18 @@ export default function QuickInserter( {
 	const onBrowseAll = () => {
 		setInserterIsOpened( {
-			rootClientId,
-			insertionIndex,
 			filterValue,
 			onSelect,
 		} );
+		setInserterInsertionPoint( { rootClientId, insertionIndex } );
```
We could have the editor stores `setInserterIsOpened` dispatch the `setInserterInsertionPoint` to the block editor store as part of the action. That said it might be clearer this way as the "set inserter is opened" communicates "toggling" rather than "setting an insertion point".

scruffian replied:
I think we might want to `setInserterInsertionPoint` without `setInserterIsOpened` as well - for example if you're using the quick inserter.

getdave replied:
But what about backwards compat on the `setInserterIsOpened` API?

scruffian replied:
What I mean is that I think its good to have the flexibility of calling these independently rather than tying one to the other.

getdave commented on packages/editor/src/store/actions.js:
```diff
@@ -720,12 +720,14 @@ export function removeEditorPanel( panelName ) {
- *                                              opened (true) or closed (false).
- *                                              To specify an insertion point,
- *                                              use an object.
- * @param {string}         value.rootClientId   The root client ID to insert at.
- * @param {number}         value.insertionIndex The index to insert at.
```
We could continue to make this backwards compatible by dispatching the new action to the block editor store _if_ the `rootClientId` or the `insertionIndex` are provided as part of `value`. You could also fire a deprecated notice in the console.

jeryj replied:
I like this idea.

getdave replied:
This API has changed. It's a public API. We have to ensure it's backwards compatible.

jeryj replied:
You can still pass these things, they just don't do anything within our implementation of the block editor. I'll add a deprecation and fire that though.

getdave replied:
> ...they just don't do anything

That was the bit I was concerned about because if a public API used to do something it needs to continue to do that. 

Looks like you fixed it up now 👍

Mamaduka replied:
We can’t deprecate a public API/feature and suggest to use a private API instead.

It would be non-actionable for consumers.

jeryj replied:
I think it would also be fine to _not_ deprecate any of it. You can still pass it those props, it just won't do anything in our codebase 🤷🏻

getdave replied:
Yeh actually we don't need to deprecate it do we? @Mamaduka is right that we can't point to a private API - great spot there.

getdave reviewed (commented):
Thanks for taking this one. Some things I noticed.

Mamaduka commented:
I have a couple of surface notes:

* Opening the inserter and setting insertion points actions are already a bit complicated. I am hesitant about introducing yet another concept in this area. I think @jeryj also noted this somewhere.
* I would like to add e2e test coverage for regressions linked to the [original PR](https://github.com/WordPress/gutenberg/pull/64048). I'll try to do this in a separate PR and close out those issues.


## Screenshot
![CleanShot 2024-09-11 at 12 46 42](https://github.com/user-attachments/assets/668a60ec-094d-42fb-8079-afd5bccf599c)

scruffian commented:
From memory, there's a difference between `blockInsertionPoint` and `insertionPoint`, and we should be using `blockInsertionPoint`, see https://github.com/WordPress/gutenberg/pull/63934

jeryj commented:
@Mamaduka 

> Opening the inserter and setting insertion points actions are already a bit complicated. I am hesitant about introducing yet another concept in this area. 

@Mamaduka - I agree, but I don't see another way to address this without blurring the boundaries of the block-editor and editor package further, as we really need `SELECT_BLOCK` to clear the insertion point (which is stored in the editor package on trunk). At the moment, the `block-editor` is emitting actions to the editor about the insertion point which then trickles back down to the `block-editor`. So, I don't think we're introducing another concept but separating out two concerns. 

I think we can make this simpler by improving the naming and docs around it. For example, it's already confusing that `setInserterIsOpen` manages the insertion point state.

jeryj committed:
Use inserterInsertionPoint for zoom out inserters and zoom out separator

jeryj commented:
@kevin940726 - Thanks for catching that! The zoom out mode inserters and zoom out mode separators were relying on `getBlockInsertionPoint` instead of the new `getInserterInsertionPoint`. Updating those fixed the zoom out separator being rendered even though there was a block selection elsewhere. Looking into the performance issue.

jeryj committed:
Remove blockInsertionPointVisible from zoom out mode inserters

jeryj committed:
Fix dragging vertical displacement on insertion points

jeryj committed:
Fix inserting at 0 index on zoom out mode

jeryj commented:
@kevin940726 - Ok - the pattern list reloading is a deeper issue. It does happen on trunk too. I haven't found the exact reproduction route yet, but it happens when using the zoom out inserters to insert some patterns. Enter them around in various places and eventually you'll get the pattern list reloading when hovering into/out of a section.

jeryj committed:
Fix insertion point at root level

scruffian reviewed (approved):
I tried to break this and I couldn't. I think it's ready!

getdave reviewed (dismissed):
We can't land this until the changes to public APIs have been handled. From what I see they haven't been yet.

Mamaduka commented:
If the Zoom out mode stays experimental for WP 6.7, then I think we should delay merging until last Gutenberg RC is released (probably next week).

jeryj commented:
I'm very fine not implementing this change before 6.7 unless necessary :) I'd like to do a larger refactor of the insertion api, as I think it needs it. Idea and issue overview forthcoming. Draft PR here: https://github.com/WordPress/gutenberg/pull/65282

jeryj committed:
Deprecate rootClientId and insertionIndex on setIsInserterOpened

scruffian commented:
> If the Zoom out mode stays experimental for WP 6.7, then I think we should delay merging until last Gutenberg RC is released (probably next week).

Is that because there could be unexpected sideeffects?

jeryj commented on packages/block-editor/src/store/private-selectors.js:
```diff
@@ -550,3 +550,13 @@ export function isZoomOutMode( state ) {
+ * @param {Object} state
+ * @return {Object} Of where the insertion point in the block editor is or null if none is set.
+ */
+export function getInserterInsertionPoint( state ) {
```
Could we name this `getInsertionPoint` instead? The same name is in the editor package. I would much prefer `getInsertionPoint` as it could be set without the inserter and we don't really want the block-editor to know about the editor package.

Mamaduka replied:
If there's no name clash, then `set/getInsertionPoint` sounds much better.

jeryj replied:
There isn't a name clash if I change enough names 😉 😅 I changed quite a bit of naming in https://github.com/WordPress/gutenberg/pull/65098/commits/64dcd1fc99efe108ccf3564dd5e178b71daa8eb1. The main source of confusion for it will be the naming of `getBlockInsertionPoint`, but at least now when you look at it it's more obvious that it's returning the state of `insertionCue`. So, while still confusing, maybe it'll be easier to understand the flow.

Mamaduka commented:
> Is that because there could be unexpected sideeffects?

@scruffian, yes. We can be more cautious if a feature/fix isn't required for the next major release.

jeryj committed:
Renaming Insertion API for clarity

- old block-editor insertionPoint becomes insertionCue (for the reducer and state value)
- which opens up insertionPoint to be the real insertion point
- so we can change getInserterInsertionPoint to getInsertionPoint
- rename editor private getInsertionPoint to getInserter to more accurately describe its state
- remove unnecessary deprecation

jeryj committed:
Catch missed renamings

jeryj committed:
More renamings and clarification

jeryj commented on packages/block-editor/src/components/block-list/zoom-out-separator.js:
```diff
@@ -29,14 +29,16 @@ export function ZoomOutSeparator( {
+		blockInsertionPoint,
 	} = useSelect( ( select ) => {
 		const {
-			getBlockInsertionPoint,
```
This is partly a bug fix, as this should not have been using `getBlockInsertionPoint`.

jeryj commented on docs/reference-guides/data/data-core-editor.md:
```diff
@@ -1420,8 +1420,12 @@ Returns an action object used to open/close the inserter.
--   _value.rootClientId_ `string`: The root client ID to insert at.
--   _value.insertionIndex_ `number`: The index to insert at.
+-   _value.rootClientId_ `string`: Deprecated. The root client ID to insert at.
+-   _value.insertionIndex_ `number`: Deprecated. The index to insert at.
```
These are not really deprecated, we're just not using them in the codebase to open the inserter anymore. You are still allowed to set these things and pass them to the inserter. I think I should undo this "Deprecated." addition here.

ajlende commented on packages/block-editor/src/store/test/private-actions.js:
```diff
@@ -123,4 +124,18 @@ describe( 'private actions', () => {
 		} );
 	} );
+
+	describe( 'setInsertionPoint', () => {
+		it( 'should return the SET_INSERTION_POINT action', () => {
+			expect(
+				setInsertionPoint( {
+					rootClientId: '',
+					index: '123',
+				} )
+			).toEqual( {
+				type: 'SET_INSERTION_POINT',
+				value: { rootClientId: '', index: '123' },
+			} );
+		} );
+	} );
```
This test and all the other tests in this file aren't doing anything useful.

The important thing for actions is that reducers do things based on them and the selectors read out a specific value. The parts of a store shouldn't be tested in isolation like this.

It looks like you're trying to fix the breaking change that may have happened some time earlier, but updating the test and documentation isn't the way to fix it. The prior logic needs to be restored.

If the tests were for the store as a whole, maybe this would have been caught.

Also, to be clear, I'm suggesting moving the tests, documentation, and fixing the `setIsInserterOpened` regression described in [your comment](#r1757473342) into a different PR.

jeryj commented on packages/editor/src/store/actions.js:
```diff
@@ -724,17 +724,34 @@ export function removeEditorPanel( panelName ) {
+			value.hasOwnProperty( 'rootClientId' ) &&
+			value.hasOwnProperty( 'insertionIndex' )
+		) {
+			return registry.dispatch( blockEditorStore ).setInsertionPoint( {
+				rootClientId: value.rootClientId,
+				index: value.insertionIndex,
+			} );
```
There are two bugs here:
1. It shouldn't return here, it should just dispatch it.
2. `setInsertionPoint` isn't an available action because it's in the private-actions in the block editor store. Are we allowed to use it in this editor package if it's private in the block editor package?

jeryj committed:
Fix logic and missing private function in editor store

jeryj committed:
Fallback to using setInserterIsOpened if already in use

jeryj committed:
Move actions test within reducer to actions test and test with selectors as well

jeryj committed:
Remove unncessary return on dispatch

jeryj commented on packages/editor/src/store/actions.js:
```diff
@@ -747,7 +747,7 @@ export const setIsInserterOpened =
 		}
 
-		dispatch( {
+		return dispatch( {
```
I didn't mean to commit this part. Removed in follow-up.

jeryj committed:
Update block editor selector state tests from insertionPoint to insertionCue

ajlende reviewed (approved):
Looks good to me. There's more to do in #65290, but this is a good start.

jeryj committed:
Move code back to previous location

Mamaduka commented:
@jeryj, I appreciate all the work done here, but I think getting broader feedback from the core team would have been nice.

jeryj commented:
@Mamaduka - I'd be love to have gotten more feedback on this as well. It had been open long enough that I didn't think I was going to hear anything else, and it was also blocking follow-up work on insertion point work with the zoom out view. Very happy to iterate or follow-up with any of it! Fortunately, I ended up finding a way to not change any public APIs, so everything is possible to iterate or undo.

youknowriad commented on packages/block-editor/src/store/private-actions.js:
```diff
@@ -359,6 +359,20 @@ export function expandBlock( clientId ) {
+ *
+ * @return {Object} Action object.
+ */
+export function setInsertionPoint( value ) {
```
How is this different than `showInsertionPoint`?

youknowriad replied:
Ok, so If I'm understand properly we have:

- `showInsertionPoint` which is the exact location where the block will be inserted. So the true insertion point.
- and we have `setInsertionPoint` which is about the top level "inserter" component. IT's about setting the "default" insertion point (if blocks are not allowed to be inserted there, we compute the true insertion point and show it using  showInsertionPoint)

is that correct?

youknowriad replied:
Clarifying my understanding more:

The "inserter insertion point" state was being set in the "editor" package because the "Inserter" component has a "rootClientId" and "clientId" props, so this one is controlled externally, I guess the idea is that at the same time, you can potentially render different inserters with different default insertion points. The current PR seems to move the state to "block-editor" but the props remain so it's still "externally" controlled right? So why the move to "block-editor" state.

jeryj replied:
> How is this different than showInsertionPoint?
`showInsertionPoint` makes the insertion cue visible. It is not code-wise related to the insertion point, just a visual indicator.

> showInsertionPoint which is the exact location where the block will be inserted. So the true insertion point.

No, this only shows where the blue line will get shown (what I will now refer to as the insertion _cue_ for clarity). We don't have a true insertion point at the moment other than passing an insertion point to `useInsertionPoint` (what the sidebar inserter was doing) or via the new `setInsertionPoint` added in this PR.

> The current PR seems to move the state to "block-editor" but the props remain so it's still "externally" controlled right? So why the move to "block-editor" state.

Exactly. No props were removed and a manually passed insertion point to the inserter will still be respected. We needed to move it to the block-editor state to clear the insertion point when a block is selected. 

[This reverted PR demos more about why we need this](https://github.com/WordPress/gutenberg/pull/64048). Code-wise, the inserter was built with a static insertion point in mind. Now the inserter can be open with selection on the block canvas happening, meaning the insertion point can be updated while the inserter is still open.

The problem with the PR above is that it relied on `getBlockInsertionPoint` which updates constantly on hover, so we had [to revert it](https://github.com/WordPress/gutenberg/pull/65208).

I wrote an overview of the Insertion API current state here: https://github.com/WordPress/gutenberg/issues/65290

youknowriad commented on packages/block-editor/src/store/private-selectors.js:
```diff
@@ -698,3 +698,13 @@ export function getClosestAllowedInsertionPointForPattern(
+ * @param {Object} state
+ * @return {Object} where the insertion point in the block editor is or null if none is set.
+ */
+export function getInsertionPoint( state ) {
```
How is this different than `getBlockInsertionPoint`?

jeryj replied:
`getBlockInsertionPoint` is really about the blue insertion cue that gets shown. `getBlockInsertionPoint` is constantly updated based on where you're hovering the canvas. Maybe originally `getBlockInsertionPoint` was about where the insertion would happen, but its code now is tied to where the insertion cue shows.

youknowriad replied:
What's the difference for you between "insertion cue" and "insertion position". If we're showing a different insertion cue than the actual insertion point, then we have problem.

draganescu replied:
The difference is that one is an index and the other one is an UI element?

youknowriad replied:
Both are state for me. The UI element is just a reflexion of a state.

jeryj replied:
I think it gets tricky when you include drag and drop. You’re showing an insertion cue at a location that may be different than what is set via selection. 

If you update that point on hover via drag and drop, this is the correct behavior for the insertion cue/insertion point, but a lot of the things built off using it were not intended to be updating their insertion point that often. 

So, I think insertion cue of where insertion will happen and should rely on insertion point if set, but the insertion cue should be able to show separate from the insertion point on instances like drag and drop. Hovering the canvas should not be updating the insertion point, it should be updating a cue about where you might be able to insert.

youknowriad replied:
> So, I think insertion cue of where insertion will happen and should rely on insertion point if set, but the insertion cue should be able to show separate from the insertion point on instances like drag and drop. Hovering the canvas should not be updating the insertion point, it should be updating a cue about where you might be able to insert.

I agree with this, but it doesn't really explain the PR to me. we were using `showInsertionPoint` for me before this PR to do this and this was already different than the "default" insertion point used in the inserter component. 

So I'm still not clear what this PR is trying to accomplish aside moving things around and potentially making them more complicated.

youknowriad commented:
One thing I'm not sure I understand is how is this blocking things in zoom-out?

jeryj commented:
> One thing I'm not sure I understand is how is this blocking things in zoom-out?

We need to be able to update the insertion point with the inserter panel still open. This isn't zoom-out specific, as it's useful for the main canvas as well. We also have work on the zoom out separators that is behaving in a buggy way, and being able to get a clear insertion point will be very helpful.

youknowriad commented:
> We need to be able to update the insertion point with the inserter panel still open. 

Can you expand more, I'm not sure how the current change is related to that. Because the previous `getInsertionPoint` or the current `getInserter` can also change their values to adapt to that (the inserter open or closed doesn't really any impact here)

jeryj commented:
> Can you expand more, I'm not sure how the current change is related to that. Because the previous `getInsertionPoint` or the current `getInserter` can also change their values to adapt to that (the inserter open or closed doesn't really any impact here)

You could change the insertion point via the old `getInsertionPoint` in the `editor` package, but, AFAIK, would have needed to change it based on the `SELECT_BLOCK` action in the `block-editor` package. I am not aware of an acceptable way to update the `editor` package via an action from the `block-editor`, and, regardless, it seems like insertion point information seems like data the `block-editor` should be in control of instead of the `editor`.

There very well may be better way to handle this, and I'm very open to that. The way I went with here seemed like the best first step towards clarifying and increasing flexibility of the insertion methods.

youknowriad commented:
> You could change the insertion point via the old getInsertionPoint in the editor package, but, AFAIK, would have needed to change it based on the SELECT_BLOCK action in the block-editor package. I am not aware of an acceptable way to update the editor package via an action from the block-editor, and, regardless, it seems like insertion point information seems like data the block-editor should be in control of instead of the editor.

There are two ways to look at it and both are valid:

 1- We want for the "block-editor" itself to control the state "default insertion point" of the global inserter. That is definitely a valid approach, in which case, you'd need a selector like this PR adds and also you'd need to use that selector consistently and remove the `rootClientId` and `clientId` props of the Inserter component (because you shouldn't be controlling the insertion point from two different and conflicting places), in which case, if something external (editor package for instance), wants to change the insertion point, it does so by calling the action `setInsertionPoint` or something on the block-editor.
 2- The default insertion point of the `Inserter` component is not controlled by a state value, instead it's just a prop that the person composing the editor (or the external packages using `Inserter`) provide as props to the `Inserter` component.


Before this PR, we used to have **2**. With this PR, we tried to move towards **1** but instead we endup with a mix of both. In my opinion, this PR is creating more confusion than solving it. We still didn't achieve 1 fully and IMO we have no reason to move towards **1** at the moment. (adding selectors, changing APIs...). I'm not sure I see a "conceptual" bug in approach 2 that forces us to reconsider it and move towards approach 1.

draganescu commented:
@youknowriad @jeryj isn't this part of Open
[Refactor Insertion API](https://github.com/WordPress/gutenberg/issues/65290)
#65290 ?

jeryj commented:
@youknowriad - without moving the insertion point to the block-editor, how would you recommend solving the [issue of blocks being inserted in the wrong place](https://github.com/WordPress/gutenberg/issues/63866) when selecting a different block? This was the cleanest way I could see to do it, but I’m very open to there being a better way. 

@draganescu - moving the conversation to the issue be great, and it does move things forwards.

youknowriad commented:
> without moving the insertion point to the block-editor, how would you recommend solving the https://github.com/WordPress/gutenberg/issues/63866 when selecting a different block? This was the cleanest way I could see to do it, but I’m very open to there being a better way.

Can you clarify which inserter is that? I don't know zoom-out mode enough.

youknowriad commented:
@jeryj Thanks for sharing the video, and the issue, this is a more convincing story and I do see why you decided to move the state to block-editor to achieve that second flow.

That said, I wonder if actually both approaches are flawed here :). Let me explain:

Previously, the inserter was open like a modal and closes when you insert, which means, if you open it using `showInsertionPoint` or something, you're certain that any other unrelated change (like selected block) won't have an impact on the insertion position because that would actually close the inserter.

Now, in the second video, you're trying to argue that now that the inserter stays open, if you select another block, you should also change the "insertion point" of the inserter. But for me there's an ambiguity here. When I click the in-between inserter and "browse all", that block is not necessarily selected but the inserter still insert into that position that differs from the selected block. 

Here's my take: what if clicking the "in-between inserter" actually selected the block instead of "setting the insertion point". For me that would be the best of two words:

 - Insertion point is not a state anymore and is always synced with the selected block.
 - no ambiguity or confusion.

Am I missing anything?

jeryj commented:
> Here's my take: what if clicking the "in-between inserter" actually selected the block instead of "setting the insertion point". For me that would be the best of two words:

While that would simplify the code, I have some concerns about unintentional side effects. The inserter responds to the selected block to filter results and allowed blocks, etc. However, maybe it all works well and simplifies it! I'll open up a PR and we can find out :)

jeryj commented:
I don't think we can rely on only block selection because of instances where we could validly insert inside of or after a selected block. The group block is the best example of this. We need to be able to insert inside of and after it.

If I use the group block inner appender, block selection would land on the group block.
If I use an inline inserter after a group block, block selection would land on the group block.

One is intended to insert _inside_ of the group and one is intended to insert _after_ the group:

| Inner group appender | in-between inserter |
|--------|--------|
| <img width="972" alt="Block canvas with quick inserter open from an inner group appender" src="https://github.com/user-attachments/assets/90ac0788-5c46-47e9-ad45-56f0ccb80833">| <img width="955" alt="Block canvas with quick inserter open from the inline appender after the group" src="https://github.com/user-attachments/assets/27054963-0ecd-43b5-90ae-f82792751508"> | 

In the above two scenarios as the codebase currently is, I don't think we could infer where to insert the block based only on block selection. The only way I could see this working via block selection is if we changed the insertion process to:

- Press the appender
- Temporarily insert a default block to this point
- Select this block
- Insertion replaces the block
- If no insertion happens or block selection changes, remove the temporary block

I think that might be similar to how the navigation block works, but I don't see that as simpler. Lmk if you see a different route here!

youknowriad commented:
you're right in the case of the appender within an empty container block, there's no easy way to translate "selection" to "insertion point". 

It's weird though, because let's say you clicked an "in between inserter" -> "browse all" to open the inserter and then you went on with your life and clicked around while the inserter is open. It's very unclear where should the block be inserted in that case.

jeryj commented:
> It's very unclear where should the block be inserted in that case.

I think bringing over the idea of the zoom out separators to the non-zoomed out view could work.

<img width="1361" alt="zoom out separator on pattern tab of block canvas showing where the patterns will be inserted" src="https://github.com/user-attachments/assets/cce57026-7287-48e1-92d0-872cf1d3607b">
